pkg/traceql/ast.go

```diff
@@ -250,27 +250,29 @@ func newSpansetOperation(op Operator, lhs SpansetExpression, rhs SpansetExpressi
 func (SpansetOperation) __spansetExpression() {}
 
 type SpansetFilter struct {
-	Expression FieldExpression
+	Expression          FieldExpression
+	matchingSpansBuffer []Span
 }
 
-func newSpansetFilter(e FieldExpression) SpansetFilter {
-	return SpansetFilter{
+func newSpansetFilter(e FieldExpression) *SpansetFilter {
+	return &SpansetFilter{
 		Expression: e,
 	}
 }
 
 // nolint: revive
-func (SpansetFilter) __spansetExpression() {}
+func (_ *SpansetFilter) __spansetExpression() {}
 
-func (f SpansetFilter) evaluate(input []*Spanset) ([]*Spanset, error) {
+func (f *SpansetFilter) evaluate(input []*Spanset) ([]*Spanset, error) {
 	var output []*Spanset
 
 	for _, ss := range input {
 		if len(ss.Spans) == 0 {
 			continue
 		}
 
-		var matchingSpans []Span
+		f.matchingSpansBuffer = f.matchingSpansBuffer[:0]
```

Comment:
we can never make the fetch/eval pipeline async. i actually love this assumption b/c it does allow us to write nice perf improvements like this. in the future the best we can do is split a pipeline by row group and execute independent pipelines in parallel

Reply:
Agree the high level parallelization has been working well and is the bulk of the performance.  The iterator nature of  Fetch I think is the main thing that ensures the engine stays single-threaded, so this optimization doesn't seem so brittle.